### ISSUE TITLE
Lint warning about self parameters

### ISSUE THREAD

**Issue by mrmonday** [Manishearth/rust-clippy#119]
Just an idea based on an issue someone had on IRC a few days back - I don't know how plausible it would be/how many false positives would occur. Take the following code:

```rust
#[derive(Copy, Clone)]
struct Foo {
    a: u32,
}

impl Foo {
    fn do_something(mut self) {
        self.a = 4;
        println!("do_something: {}", self.a);
    }
    
    fn print(self) {
        println!("print: {}", self.a);
    }
}

fn main() {
    let val = Foo { a: 0 };
    
    val.do_something();
    
    val.print();
}
```

Playpen: https://play.rust-lang.org/?gist=69f8e08682f48177f43c&version=stable

Here the user wrote this code, and likely encountered compile errors about `val` being moved, and followed suggestions to impl `Copy` and `Clone` for `Foo`. This caused the code to compile, however it produced unexpected results (the user expected 4 to be printed both times). The solution of course is for both methods to borrow self, rather than taking ownership.

I suggest a lint which checks the self parameter of methods (or, perhaps, any parameters for any functions), and checks that they take the minimal possible amount of ownership, ie: prefer no self parameter to &self, &self to &mut self, self to mut self etc. There are obviously lots of potential constraints here, eg if you're implementing a trait, or future proofing a method then you won't want an error to occur, but it could be useful in scenarios like the above, where the user is fighting the compiler to get code compiling.

**Comment by llogiq** [comment]
There already is a lint that warns on unneeded mutable borrowing.

However, there are valid reasons to consume self by value – e.g. `Drain impl`s. So that lint cannot decide whether a match was wrong or not.

A possible approach is to check the function name for keywords, e.g. if it contains 'drain' 'drop' or 'into', taking `self` by value won't trigger a warning.

**Comment by mrmonday** [comment]
That's another case I thought of, but neglected to mention. Obviously it's a tough task to figure out if consumption was desired/necessary, keywords could be the right way to go.

In this particular case it could be possible to identify usage patterns, eg, "struct impls Copy, and all methods take ownership - did you mean to use move semantics and use &self/&mut self instead?", or perhaps "structure is mutated, but is dropped at the end of the method do_something() - did you mean to use &mut self?". There are probably a few other cases which could be investigated here, without introducing lots of false positives when dropping was actually desired.

**Comment by Manishearth** [comment]
I think this is something that `rustc` diagnostics should improve, not clippy. Besides, lints are run after borrowck so this won't be helpful in most cases and will have lots of false positive anyway.

**Comment by llogiq** [comment]
Agreed. Perhaps borrowck could be extended to hint that a function could be changed not to consume self to fix a problem *if* it would make the problem go away without introducing errors elsewhere.

However, I think that @mrmonday should perhaps just blog about it – once people know about the issue, it's easy enough to work around.

**Comment by Manishearth** [comment]
Closing, since lints run after borrowck
